core/store/src/db/rocksdb/snapshot.rs

```diff
@@ -0,0 +1,182 @@
+///
+/// Serves as kind of RAII type which logs information about the checkpoint when
+/// object is dropped if the checkpoint hasn’t been removed beforehand by
+/// [`Snapshot::remove`] method.
+///
+/// The usage of the type is to create a checkpoint before a risky database
+/// operation and then [`Snapshot::remove`] it when the operation finishes
+/// successfully.  In that scenario, the checkpoint will be gracefully deleted
+/// from the file system.  On the other hand, if the operation fails an the
+/// checkpoint is not deleted, this type’s Drop implementation will log
+/// informational messages pointing where the snapshot resides and how to
+/// recover data from it.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct Snapshot(std::path::PathBuf);
+
+/// Possible errors when creating a checkpoint.
+#[derive(Debug)]
+pub enum SnapshotError {
+    /// Snapshot at requested location already exists.
+    ///
+    /// More specifically, the specified path exists (since the code does only
+    /// rudimentary check whether the path is a checkpoint or not).
+    AlreadyExists(std::path::PathBuf),
+
+    /// Error while creating a snapshot.
+    IOError(io::Error),
+}
+
+impl std::convert::From<io::Error> for SnapshotError {
+    fn from(err: io::Error) -> Self {
+        Self::IOError(err)
+    }
+}
+
+impl std::convert::From<::rocksdb::Error> for SnapshotError {
+    fn from(err: ::rocksdb::Error) -> Self {
+        super::into_other(err).into()
+    }
+}
+
+impl Snapshot {
+    /// Creates a new snapshot for given database.
+    ///
+    /// If the snapshot already exists, returns [`SnapshotError::AlreadyExists`]
+    /// error.  (More specifically, if the `snapshot_path` already exists since
+    /// the method does not make any more sophisticated checks whether the path
+    /// contains a snapshot or not).
+    pub(crate) fn new(
+        db_path: &std::path::Path,
+        config: &crate::StoreConfig,
+        snapshot_path: std::path::PathBuf,
+    ) -> Result<Self, SnapshotError> {
+        tracing::info!(target: "db", snapshot_path=%snapshot_path.display(),
+                       "Creating database snapshot");
+        if snapshot_path.exists() {
+            return Err(SnapshotError::AlreadyExists(snapshot_path));
+        }
+
+        let db = super::RocksDB::open(db_path, config, crate::Mode::ReadWrite)?;
+        let cp = Checkpoint::new(&db.db).map_err(super::into_other)?;
+        cp.create_checkpoint(&snapshot_path)?;
+
+        Ok(Self(snapshot_path))
+    }
+
+    /// Deletes the checkpoint from the file system.
+    ///
+    /// If the deletion fails, error is logged but the function does not fail.
+    pub fn remove(self) {
+        // SAFETY: Self is a repr(transparent) containing PathBuf so this
+        // transmute is guaranteed to be safe.  We’re doing this to avoid
+        // Snapshot::drop call.
+        let path: std::path::PathBuf = unsafe { core::mem::transmute(self) };
```

Comment:
That seems a bit overcomplicated. Wouldn't it be simpler to use a `bool` flag as a field of `Snapshot`, set it upon successful removal of the snapshot and then conditionally log in the drop implementation?

Reply:
I'd write this as 

```rust
    pub fn remove(self) {
        let mut me = std::mem::ManuallyDrop::new(self);
        let path = std::mem::take(&mut me.0);
```

Reply:
I thought about adding a flag or using `Option<PathBuf>` but it seems to me like a hack.  The struct doesn’t have two states.  It has a single state: there’s a snapshot and it holds path to it.  `remove` method is one way in which the object can be consumed and dropped.

Reply:
You don't need `Option`, `PathBuf` is already essentially an `Option` -- similar to String, Vec and other collections, it's `Default` empty state doesn't allocate, and doesn't need to be dropped.

Reply:
Yes, but I still need a flag in Snapshot which tells me whether the snapshot exists and PathBuf stores path to it or it doesn’t.

Reply:
That's where `ManuallyDrop` comes from -- if you are running drop , you know you should warn.

Reply:
Right, that’s my original approach.  I’ve since changed the code in a way that @jakmeier suggested.

Reply:
Yeah `ManuallyDrop` would also be nicer than transmuting. But personally, I like the flag better and if feels lees hacky to me than juggling with conditions to call or not call `drop`. From the compiler's point of view, this struct has two states. And depending on which one it is in, it will have to do different things when it goes out of scope.